return type 'null' not allowed

### Zig Version

0.11.0-dev.3620+c76ce25a6

### Steps to Reproduce and Observed Behavior

```zig
const std = @import("std");

test {
    try std.testing.expect(foo() == null);
}

fn foo() @Type(.Null) {
    return null;
}
```

```
test.zig:7:10: error: return type '@TypeOf(null)' not allowed
fn foo() @Type(.Null) {
         ^~~~~~~~~~~~
referenced by:
    test_0: test.zig:4:28
```

### Expected Behavior

null literal should behave as any other zero-sized type. I expect all 1 tests passed

> null literal should behave as any other zero-sized type

I think that means this should all pass too, just like `u0` and `void`.:

```zig
test {
    var x: @TypeOf(null) = null;
    x = null;

    foo(&x);

    @compileLog(@sizeOf(@TypeOf(null)));
    @compileLog(@sizeOf(@Type(.Null)));
}

fn foo(x: *@TypeOf(null)) void {
    x.* = null;
}
```

Currently:
```
x.zig:4:12: error: variable of type '@TypeOf(null)' must be const or comptime
    var x: @TypeOf(null) = null;
           ^~~~~~~~~~~~~
x.zig:7:25: error: no size available for type '@TypeOf(null)'
    @compileLog(@sizeOf(@TypeOf(null))); // should be 0
                        ^~~~~~~~~~~~~
x.zig:13:8: error: parameter of type '*@TypeOf(null)' must be declared comptime
fn foo(x: *@TypeOf(null)) void {
       ^~~~~~~~~~~~~~~~~
```

Some relevant code is probably https://github.com/ziglang/zig/blob/423d7b848b1953173df99fde1f83166dc68c2a2c/src/type.zig#L2404-L2416

--- a/src/type.zig
+++ b/src/type.zig
@@ -2403,14 +2403,14 @@ pub const Type = struct {
 
     pub fn isValidParamType(self: Type, mod: *const Module) bool {
         return switch (self.zigTypeTagOrPoison(mod) catch return true) {
-            .Undefined, .Null, .Opaque, .NoReturn => false,
+            .Opaque, .NoReturn => false,
             else => true,
         };
     }
 
     pub fn isValidReturnType(self: Type, mod: *const Module) bool {
         return switch (self.zigTypeTagOrPoison(mod) catch return true) {
-            .Undefined, .Null, .Opaque => false,
+            .Opaque => false,
             else => true,
         };
     }

--- a/test/behavior/fn.zig
+++ b/test/behavior/fn.zig
@@ -580,3 +580,17 @@ test "lazy values passed to anytype parameter" {
     const D = struct {};
     try expect(@sizeOf(D) << 1 == 0);
 }
+
+test "pass and return comptime-only types" {
+    const S = struct {
+        fn returnNull(comptime x: @Type(.Null)) @Type(.Null) {
+            return x;
+        }
+        fn returnUndefined(comptime x: @Type(.Undefined)) @Type(.Undefined) {
+            return x;
+        }
+    };
+
+    try expectEqual(null, S.returnNull(null));
+    try expectEqual(@as(u0, 0), S.returnUndefined(undefined));
+}

--- a/test/cases/compile_errors/function_parameter_is_opaque.zig
+++ b/test/cases/compile_errors/function_parameter_is_opaque.zig
@@ -4,23 +4,11 @@ export fn entry1() void {
     _ = someFuncPtr;
 }
 
-export fn entry2() void {
-    const someFuncPtr: fn (@TypeOf(null)) void = undefined;
-    _ = someFuncPtr;
-}
-
 fn foo(p: FooType) void {
     _ = p;
 }
 export fn entry3() void {
     _ = foo;
-}
-
-fn bar(p: @TypeOf(null)) void {
-    _ = p;
-}
-export fn entry4() void {
-    _ = bar;
 }
 
 // error
@@ -29,7 +17,5 @@ export fn entry4() void {
 //
 // :3:28: error: parameter of opaque type 'tmp.FooType' not allowed
 // :1:17: note: opaque declared here
-// :8:28: error: parameter of type '@TypeOf(null)' not allowed
-// :12:8: error: parameter of opaque type 'tmp.FooType' not allowed
+// :7:8: error: parameter of opaque type 'tmp.FooType' not allowed
 // :1:17: note: opaque declared here
-// :19:8: error: parameter of type '@TypeOf(null)' not allowed

--- a/test/cases/compile_errors/function_returning_opaque_type.zig
+++ b/test/cases/compile_errors/function_returning_opaque_type.zig
@@ -2,12 +2,6 @@ const FooType = opaque {};
 export fn bar() FooType {
     return error.InvalidValue;
 }
-export fn bav() @TypeOf(null) {
-    return error.InvalidValue;
-}
-export fn baz() @TypeOf(undefined) {
-    return error.InvalidValue;
-}
 
 // error
 // backend=stage2
@@ -15,5 +9,3 @@ export fn baz() @TypeOf(undefined) {
 //
 // :2:17: error: opaque return type 'tmp.FooType' not allowed
 // :1:17: note: opaque declared here
-// :5:17: error: return type '@TypeOf(null)' not allowed
-// :8:17: error: return type '@TypeOf(undefined)' not allowed

